solutions/advanced/spoj-cot.mdx

```diff
@@ -0,0 +1,155 @@
+---
+id: spoj-cot
+source: SPOJ
+title: Count on a tree
+author: Nigar Hajiyeva
+---
+## Explanation
+
+
+In this code, a persistent segment tree is built for each node in the tree by creating a new version based on its parent’s segment tree.
+
+First, a persistent segment tree for the root node which shows the frequency of the values encountered is built. We set all of the values of this tree to zeros. To incorporate the value of the current node, we create a new version of the segment tree by adding nodes to the parent’s tree and incrementing the segment tree node corresponding to the current node’s value. We store the root node of the new version of the tree for future reference. This way we don't need to build a completely new tree and modify only the necessary nodes while reusing the unchanged parts.
+
+Each persistent segment tree keeps the frequency of numbers on the way from the root of the tree given in the input to itself. When asked for the kth minimum node in the path of $u$ to $v$, the code first determines the lowest common ancestor (LCA) of the tree in the input using binary lifting. Next, we run a binary search using the frequency values from the persistent segment tree versions of four nodes: $u$, $v$, the LCA, and the parent of the LCA.
+
+Since the values that are not on the path of $u$ and $v$, but on the path from the root node to the nodes $u$ and $v$ are also counted in their persistent segment trees, just considering their persistent segment trees would produce the wrong results.
+
+![example](spoj-cot/example.png)
+
+**Handling Overcounting with LCA**
+
+We subtract the frequencies of nodes from the root node to the LCA and its parent because we count the nodes twice from the root node to the LCA. In addition, since the LCA is on the path of $u$ to $v$, we subtract the frequencies from the LCA and the parent of the LCA to consider the value of the LCA.
+
+$$
+\texttt{leftCountFromUToV} = \texttt{leftCount}[\texttt{$u$}]+\texttt{leftCount}[\texttt{$v$}]−\texttt{leftCount}[\texttt{LCA}]−\texttt{leftCount}[\texttt{parentOfLCA}]
+$$
+
+
+
+**Note:**
+Value compression is used to effectively handle large numbers.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}((N+M) \cdot \log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+
+#include <bits/stdc++.h>
+using namespace std;
+const int MAX = 2e5, MAX_NODES = 2e6, LOG = 20;
+int L[MAX_NODES + 1], R[MAX_NODES + 1], t[MAX_NODES + 1], roots[MAX + 1],
+    up[MAX + 1][LOG + 1];
+int tin[MAX + 1], tout[MAX + 1], val[MAX + 1];
+vector<int> graph[MAX + 1];
+int nxt = 1, timer = 0;
+
+int update(int v, int l, int r, int pos) {
```

Comment:
are these functions all one giant data structure or are there multiple data structures? if possible i'd put them inside a struct so readers can see how exactly you're interfacing with the numbers